src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAwaitForeachTests.cs

```diff
@@ -4237,6 +4237,95 @@ public IAsyncEnumerator<int> GetAsyncEnumerator(System.Threading.CancellationTok
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+class C
+{
+    public static async Task M(IAsyncEnumerable<int> e)
+    {
+        await foreach (var i in e)
+        {
+        }
+    }
+}
+namespace System.Collections.Generic
+{
+    public interface IAsyncEnumerable<out T>
+    {
+        IAsyncEnumerator<T> GetAsyncEnumerator(System.Threading.CancellationToken token);
+    }
+
+    public interface IAsyncEnumerator<out T> : System.IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask<bool> MoveNextAsync();
+        T Current { get; }
+    }
+}
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(
+                // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'IAsyncEnumerable<int>' because 'IAsyncEnumerable<int>' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                //         await foreach (var i in e)
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "e").WithArguments("System.Collections.Generic.IAsyncEnumerable<int>", "GetAsyncEnumerator").WithLocation(8, 33)
+                );
+        }
+
+        [Fact]
+        public void GetAsyncEnumerator_CancellationTokenMustBeOptional_OnIAsyncEnumerable_ExplicitImplementation()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Threading.Tasks;
+class C : IAsyncEnumerable<int>
+{
+    public static async Task M(C c)
+    {
+        await foreach (var i in c)
+        {
+        }
+    }
+    IAsyncEnumerator<int> IAsyncEnumerable<int>.GetAsyncEnumerator(System.Threading.CancellationToken token) => throw null;
+}
+namespace System.Collections.Generic
+{
+    public interface IAsyncEnumerable<out T>
+    {
+        IAsyncEnumerator<T> GetAsyncEnumerator(System.Threading.CancellationToken token);
+    }
+
+    public interface IAsyncEnumerator<out T> : System.IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask<bool> MoveNextAsync();
+        T Current { get; }
+    }
+}
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+";
+            var comp = CreateCompilationWithTasksExtensions(source);
+            comp.VerifyDiagnostics(
+                // (8,33): error CS8411: Asynchronous foreach statement cannot operate on variables of type 'C' because 'C' does not contain a suitable public instance or extension definition for 'GetAsyncEnumerator'
+                //         await foreach (var i in c)
+                Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "c").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33)
```

Comment:
>Diagnostic(ErrorCode.ERR_AwaitForEachMissingMember, "c").WithArguments("C", "GetAsyncEnumerator").WithLocation(8, 33) [](start = 16, length = 117)

Is this a breaking change? It looks like we simply used default value for the token in this case. In fact, I am surprised to see a difference in handling for the two scenarios (prior to the latest adjustment). It shouldn't matter whether we use the interface directly or use a type implementing the interface. Both implement the interface, an interface "implements" itself.

Reply:
So, this is a breaking change, but the code shouldn't have been allowed. The method definition does not have a default parameter, and the spec explicitly calls out that these methods must be able to called with no parameters. I can send an email to compat council if you'd like, but the number of people who would be affected by such a change is very, very small, and they have to be operating in an unsupported scenario by polyfilling these APIs anyway.

---
In reply to: [550235389](https://github.com/dotnet/roslyn/pull/50105#discussion_r550235389) [](ancestors = 550235389)

Reply:
It's also worth noting that we regressed this scenario in master: this test case will null ref in the local rewriter in master currently, because we changed the way default argument binding works and this case was missed.

---
In reply to: [550290615](https://github.com/dotnet/roslyn/pull/50105#discussion_r550290615) [](ancestors = 550290615,550235389)

Reply:
> It's also worth noting that we regressed this scenario in master: this test case will null ref in the local rewriter in master currently, because we changed the way default argument binding works and this case was missed.

I think we should restore the original behavior. The purpose of the change to move creation of default values in binding vs. lowering was a refactoring rather than an introduction of a behavior change. We don't know what else might be broken.

---
In reply to: [550291362](https://github.com/dotnet/roslyn/pull/50105#discussion_r550291362) [](ancestors = 550291362,550290615,550235389)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/50182](https://github.com/dotnet/roslyn/issues/50182) to track actually erroring in this scenario.

---
In reply to: [550331559](https://github.com/dotnet/roslyn/pull/50105#discussion_r550331559) [](ancestors = 550331559,550291362,550290615,550235389)